### PR TITLE
nitrokey-app2: init at 2.1.2

### PR DESCRIPTION
**Pull request by 999eagle**
###### Description of changes

upstream URL: https://github.com/Nitrokey/nitrokey-app2

additional python dependencies:
- https://pypi.org/project/PyQt5-stubs/
- https://pypi.org/project/qt-material/

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/security/nitrokey-app2/default.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    pynitrokey
+    pyudev
+    pyqt5
+    pyqt5-stubs
+    qt-material
+  ];
+
+  preFixup = ''
+    qtWrapperArgs+=("''${gappsWrapperArgs[@]}")
+    wrapQtApp "$out/bin/nitrokeyapp"
```
Comment (SuperSandro2000) [inline pkgs/tools/security/nitrokey-app2/default.nix:44]: Is this double wrapping?

File: pkgs/development/python-modules/pyqt5-stubs/default.nix [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+buildPythonPackage rec {
+  pname = "PyQt5-stubs";
+  version = "5.15.6.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-kScKwj6/OKHcBM2XqoUs0Ir4Lcg5EA5Tla8UR+Pplwc=";
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyqt5-stubs/default.nix:14]: Missing pythonImportsCheck

**Comment by SuperSandro2000** [comment]
Can we avoid this?

```
       error: Package ‘nrfutil-6.1.7’ in /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/ofborg-evaluator-1/pkgs/development/tools/misc/nrfutil/default.nix:49 has an unfree license (‘unfreeRedistributable’), refusing to evaluate.
       a) To temporarily allow unfree packages, you can use an environment variable
          for a single invocation of the nix tools.
            $ export NIXPKGS_ALLOW_UNFREE=1
        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.
       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnfree = true; }
       in configuration.nix to override this.
       Alternatively you can configure a predicate to allow specific packages:
         { nixpkgs.config.allowUnfreePredicate = pkg: builtins.elem (lib.getName pkg) [
             "nrfutil"
           ];
         }
       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnfree = true; }
       to ~/.config/nixpkgs/config.nix.
Cannot nix-instantiate `nitrokey-app2.passthru.tests' because:
```

File: pkgs/development/python-modules/pyqt5-stubs/default.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+}:
+
+buildPythonPackage rec {
+  pname = "PyQt5-stubs";
+  version = "5.15.6.0";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyqt5-stubs/default.nix:8]:
```suggestion
  version = "5.15.6.0";
  format = "setuptools";
  disabled = pythonOlder "3.7";
```

**Comment by 999eagle** [comment]
`nrfutil` is a dependency of `pynitrokey` and is required for the Nitrokey 3, though it seems like upstream has removed the dependency by "vendoring" (i.e. copying) the relevant source code into their source tree. So it's likely that the dependency on `nrfutil` could be removed from `pynitrokey`. I'll check whether that's possible in a second PR.

File: pkgs/tools/security/nitrokey-app2/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  preBuild = ''
+    pyrcc5 nitrokeyapp/ui/resources.qrc -o nitrokeyapp/ui/resources_rc.py
+    for ui_file in nitrokeyapp/ui/*.ui; do
+      pyuic5 --from-imports $ui_file -o ''${ui_file%ui}py
+    done
+  '';
+
+  nativeBuildInputs = with python3.pkgs; [
+    pyqt5
```
[the remark this answers is not in the document]
Comment (999eagle) [reply pkgs/tools/security/nitrokey-app2/default.nix:25]: As far as I can tell, no, we do not. `nitrokey-app2` runs and works fine without any of these features. The features there are only required for the full tests of `pyqt5-stubs`.

File: pkgs/development/python-modules/pyqt/5.x.nix [hunk 5]
```diff
@@ -117,6 +117,10 @@ buildPythonPackage rec {
     multimediaEnabled = withMultimedia;
     webKitEnabled = withWebKit;
     WebSocketsEnabled = withWebSockets;
+    connectivityEnabled = withConnectivity;
+    locationEnabled = withLocation;
+    serialPortEnabled = withSerialPort;
+    toolsEnabled = withTools;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyqt/5.x.nix:123]: We should really sort this alphabetically

File: pkgs/development/python-modules/pyqt5-stubs/default.nix [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  + lib.optionalString (!pyqt5.locationEnabled) ''
+    rm tests/qflags/test_QtLocation_*
+    rm tests/qflags/test_QtPositioning_*
+  ''
+  + lib.optionalString (!pyqt5.multimediaEnabled) ''
+    rm tests/qflags/test_QtMultimedia_*
+  ''
+  + lib.optionalString (!pyqt5.serialPortEnabled) ''
+    rm tests/qflags/test_QtSerialPort_*
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyqt5-stubs/default.nix:45]:
```suggestion

  postPatch = ''
    # pulls in a dependency to mypy, but we don't want to run linters
    rm tests/test_stubs.py
  '' + lib.optionalString (!pyqt5.connectivityEnabled) ''
    rm tests/qflags/test_QtBluetooth_*
    rm tests/qflags/test_QtNfc_*
  '' + lib.optionalString (!pyqt5.toolsEnabled) ''
    rm tests/qflags/test_QtDesigner_*
  '' + lib.optionalString (!pyqt5.locationEnabled) ''
    rm tests/qflags/test_QtLocation_*
    rm tests/qflags/test_QtPositioning_*
  '' + lib.optionalString (!pyqt5.multimediaEnabled) ''
    rm tests/qflags/test_QtMultimedia_*
  '' + lib.optionalString (!pyqt5.serialPortEnabled) ''
    rm tests/qflags/test_QtSerialPort_*
  '';
```

File: pkgs/tools/security/nitrokey-app2/default.nix [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+}:
+
+python3.pkgs.buildPythonApplication rec {
+  pname = "nitrokey-app2";
+  version = "2.1.2";
+  format = "flit";
+
+  src = fetchFromGitHub {
+    owner = "Nitrokey";
+    repo = pname;
```
Comment (SuperSandro2000) [inline pkgs/tools/security/nitrokey-app2/default.nix:15]:
```suggestion
    repo = "nitrokey-app2";
```

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/python-modules/pyqt/5.x.nix (hunk 5 above)
Comment (999eagle) [reply pkgs/development/python-modules/pyqt/5.x.nix:123]: Should we also sort the other uses of those flags for pyqt5 while we're at it?

**Comment by panicgh** [comment]
FYI #254924

File: pkgs/tools/security/nitrokey-app2/default.nix [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  preFixup = ''
+    wrapQtApp "$out/bin/nitrokeyapp" \
+      --set-default CRYPTOGRAPHY_OPENSSL_NO_LEGACY 1
+  '';
+
+  meta = with lib; {
+    description = "This application allows to manage Nitrokey 3 devices";
+    homepage = "https://github.com/Nitrokey/nitrokey-app2";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ _999eagle ];
```
Comment (mibmo) [inline pkgs/tools/security/nitrokey-app2/default.nix:48]: Set the main program so you can i.e. `nix run` it
```suggestion
    maintainers = with maintainers; [ _999eagle ];
    mainProgram = "nitrokeyapp";
```

(I'm not sure why the suggestion tool wants to delete and re-add the same line but :+1:)

**Review by mibmo — APPROVED** [review]

**Review by panicgh — APPROVED** [review]
Tested successfully with a Nitrokey 3.

Result of `nixpkgs-review pr 243049` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>nitrokey-app2</li>
    <li>nitrokey-app2.dist</li>
    <li>python310Packages.pyqt5-stubs</li>
    <li>python310Packages.pyqt5-stubs.dist</li>
    <li>python310Packages.qt-material</li>
    <li>python310Packages.qt-material.dist</li>
    <li>python311Packages.pyqt5-stubs</li>
    <li>python311Packages.pyqt5-stubs.dist</li>
    <li>python311Packages.qt-material</li>
    <li>python311Packages.qt-material.dist</li>
  </ul>
</details>

**Comment by panicgh** [comment]
@SuperSandro2000 Maybe you can have another look and maybe merge? It would be unfortunate if we wasted the work of 4 persons.

**Pull request merged**